.travis/script.sh

```diff
@@ -18,11 +18,775 @@
 
 PREV_PATH="`pwd`"
 
-die() {
+die()
+{
 	echo " *** ERROR: " $*
 	exit 1
 }
 
+#######################################
+# Prepare android build system
+# Globals:
+#   None
+# Arguments:
+#   None
+# Returns:
+#   None
+#######################################
+android_prepare_build_system()
+{
+    # Android build system
+    (mkdir build && cd build && git init && git pull --depth 1 https://android.googlesource.com/platform/build 2db32730e79cafcf13e1f898a7bee7f82b0449d6)
+    ln -s build/core/main.mk Makefile
+
+    # Workarounds for java checking
+    export ANDROID_JAVA_HOME=/usr/lib/jvm/java-8-oracle
+    mkdir bin
+    cat > bin/java <<EOF
+#!/bin/sh
+echo java version \"1.6\"
+EOF
+
+    cat > bin/javac <<EOF
+echo javac \"1.6\"
+EOF
+    chmod a+x bin/java bin/javac
+    export PATH=$(pwd)/bin:$PATH
+
+    # Files for building ndk
+    mkdir -p system/core/include/arch/linux-arm
+    touch system/core/include/arch/linux-arm/AndroidConfig.h
+
+    mkdir -p system/core/include/arch/linux-x86
+    touch system/core/include/arch/linux-x86/AndroidConfig.h
+
+    ANDROID_NDK_PATH=$(dirname $(which sdkmanager))/../../ndk-bundle
+    mkdir -p bionic/libc/
+    cp -r $ANDROID_NDK_PATH/sysroot/usr/include bionic/libc/include
+    mv bionic/libc/include/arm-linux-androideabi/asm bionic/libc/include/asm
+
+    mkdir -p out/target/product/generic/obj/
+    cp -r $ANDROID_NDK_PATH/platforms/android-27/arch-arm/usr/lib out/target/product/generic/obj/
+
+    mkdir -p bionic/libstdc++
+    cp -r $ANDROID_NDK_PATH/sources/cxx-stl/gnu-libstdc++/4.9/include bionic/libstdc++
+    cp -r $ANDROID_NDK_PATH/sources/cxx-stl/gnu-libstdc++/4.9/libs/armeabi-v7a/include/* bionic/libstdc++/include
+    # The default libstdc++.so does not contain full stl implementation, see https://developer.android.com/ndk/guides/cpp-support
+    cp -r $ANDROID_NDK_PATH/sources/cxx-stl/gnu-libstdc++/4.9/libs/armeabi-v7a/libgnustl_shared.so out/target/product/generic/obj/lib/libstdc++.so
+
+    # Build spec
+    cat > buildspec.mk <<EOF
+TARGET_PRODUCT := generic
+TARGET_BUILD_VARIANT := eng
+TARGET_BUILD_TYPE := release
+TARGET_TOOLS_PREFIX := $ANDROID_NDK_PATH/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/bin/arm-linux-androideabi-
+EOF
+}
+
+#######################################
+# Prepare dbus source code
+# Globals:
+#   None
+# Arguments:
+#   None
+# Returns:
+#   None
+#######################################
+android_prepare_dbus()
+{
+    DBUS_SOURCE=dbus-1.4.26
+    wget "https://dbus.freedesktop.org/releases/dbus/${DBUS_SOURCE}.tar.gz"
+    tar xvf "${DBUS_SOURCE}.tar.gz"
+    cat > "${DBUS_SOURCE}/config.h" <<EOF
+/* config.h.  Generated from config.h.in by configure.  */
```

Comment:
How about adding a copy of this as a file instead of embedded in the script?

Reply:
As you see there are some other files generated in this script, all of them are just used for travis test. Some of these files are tricks to cheat the android build system. I prefer keeping these files generated at runtime.